ath79: add support for Compex WPJ563

yogo1212 opened the pull request:
This is the last of the 3 compex devices i've prepared ports from ar71xx to ath79 for.
It's much like the WPJ531 but with three antenna-sockets for the on-board radio.

yogo1212 committed:
ath79: add support for Compex WPJ563

Specifications:

SoC: QCA9563
DRAM: 128MB DDR2
Flash: 16MB SPI-NOR
2 Gigabit ethernet ports
3×3 2.4GHz on-board radio
miniPCIe slot that supports 5GHz radio
PoE 24V passive or 36V-56V passive with optional IEEE 802.3af/at
USB 3.0 header

Installation:

To install, either start tftp in bin/targets/ath79/generic/ and use
the u-boot prompt over UART:

tftpboot 0x80500000 openwrt-ath79-generic-compex_wpj563-squashfs-sysupgrade.bin
erase 0x9f680000 +1
erase 0x9f030000 +$filesize
cp.b $fileaddr 0x9f030000 $filesize
boot

The cpximg file can be used with sysupgrade in the stock firmware (add
SSH key in luci for root access) or with the built-in cpximg loader.
The cpximg loader can be started either by holding the reset button
during power up or by entering the u-boot prompt and entering 'cpximg'.
Once it's running, a TFTP-server under 192.168.1.1 will accept the image
appropriate for the board revision that is etched on the board.

For example, if the board is labelled '7A02':

tftp -v -m binary 192.168.1.1 -c put openwrt-ath79-generic-compex_wpj563-squashfs-cpximg-7a02.bin

MAC addresses:

<&uboot 0x2e010>  *:71  (label)
<&uboot 0x2e018>  *:72
<&uboot 0x2e020>  *:73
<&uboot 0x2e028>  *:74

Only the first two are used (for ethernet), the WiFi modules have
separate (valid) addresses. The latter two addresses are not used.

Signed-off-by: Leon M. George <leon@georgemail.eu>

adrianschmutzler commented:
Please have a look at the style updates I did in a0bb356612fddf747893fffbfa216dcb3c05f069 (also for the commit message) and apply them here as well.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -131,7 +131,8 @@ ath79_setup_interfaces()
 		;;
-	compex,wpj344-16m)
+	compex,wpj344-16m|\
+	compex,wpj563)
```
ar71xx has 2:lan 3:wan for this one?

yogo1212 replied:
i've double-checked the port mapping using swconfig. port 3 is next to the LEDs (lan in stock), port 2 is next to the DC socket (wan in stock).

in the official openwrt-sdk (not to be mistaken with osdk), it is the other way around:
https://github.com/compex-systems/lede/blob/openwrt-18.06/target/linux/ar71xx/base-files/etc/board.d/02_network#L538

i find the stock solution (wan next to the power jack) more intuitive but i don't mind either way.

adrianschmutzler replied:
I prefer the stock solution as well. However, that means that ar71xx has the "wrong" port assignment right now?

yogo1212 replied:
"wrong" is a harsh term :-D
i've just noticed compex uses [mehlis code](https://github.com/compex-systems/lede/blame/openwrt-18.06/target/linux/ar71xx/base-files/etc/board.d/02_network#L538) for their openwrt SDK.

i'll do a third check with compexWRT/stock/OSDK (have only one board here).

adrianschmutzler replied:
But since this is just a "board", there are no external indicators and the choice of LAN vs. WAN is conceptually arbitrary in any case? So, we might just decide by ourselves what is right here, vendor firmware or previous established use in ar71xx?

adrianschmutzler replied:
Still waiting with further steps for an answer to this "question" ...

yogo1212 replied:
i'm very sorry to have kept you waiting. at one point last week i was very unsure about which firmware i tested.
the information about stock port assignment i've given above is wrong.

both ports are lan ports in stock (like with the wpj428).
i still prefer the more intuitive (current) port assignment.

tested using this firmware:
http://downloads.compex.com.sg/?dir=uploads/CompexWRT/Stable/v6.2.12-b190718-OEM

```
14: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 04:f0:21:3d:ec:71 brd ff:ff:ff:ff:ff:ff
15: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
    link/ether 04:f0:21:3d:ec:71 brd ff:ff:ff:ff:ff:ff
16: eth0.2@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DEFAULT group default qlen 1000
    link/ether 04:f0:21:3d:ec:71 brd ff:ff:ff:ff:ff:ff
```

adrianschmutzler commented on target/linux/ath79/dts/qca9563_compex_wpj563.dts:
```diff
@@ -0,0 +1,132 @@
+			label = "wpj563:green:sig4";
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+		};
+	};
```
What about the buzzer set up in ar71xx?

yogo1212 replied:
completely forgot about that :speak_no_evil:

adrianschmutzler commented on target/linux/ath79/dts/qca9563_compex_wpj563.dts:
```diff
@@ -0,0 +1,132 @@
+&mdio0 {
+	status = "okay";
+
+	phy-mask = <0>;
```
empty line after phy-mask

yogo1212 replied:
done

adrianschmutzler commented on target/linux/ath79/dts/qca9563_compex_wpj563.dts:
```diff
@@ -0,0 +1,132 @@
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&uboot 0x18>;
```
This line should be removed?

yogo1212 replied:
you're right

yogo1212 replied:
done

adrianschmutzler commented on target/linux/ath79/dts/qca9563_compex_wpj563.dts:
```diff
@@ -0,0 +1,132 @@
+		sig1 {
+			label = "wpj563:green:sig1";
+			gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
+		};
```
ar71xx sets up this led for diag. So, you could use it via the aliases node (or another one of the sigX leds dependent on which makes sense) ...

yogo1212 replied:
i've chosen sig4 to be consistent with the wpj531 and because i've noticed that sometimes sig1 is triggered by something else (don't know what).

yogo1212 replied:
haven't had sig1 do anything for a while now.
probably a mistake on my end from when i tested the numerous different vendor firmwares.


in (openwrt) ar71xx, the choice for the status LED is not consistent: wpj342, wpj531, and wpj558 use sig3; wpj563 uses sig1 .
with the compex firmware, all LEDs are used for rssiled + sig4 for "diag" (except for the wpj342).
sig4 is used by u-boot and is active when booting the stock firmware.
it's documented as 'RSSI4/Diag' in the hardware guides for WPJ344/WPJ531/WPJ563.

i'd say sig4 is the best choice for the status LED for consistency with the stock-image, the official documentation, and the preinstalled u-boot.

adrianschmutzler commented:
Looks good in general, only a few comments ...

yogo1212 commented:
i've added the `Specification:` and `Installation:` headings and changed the line about POE to `PoE 24V passive or 36V-56V passive with optional IEEE 802.3af/at`.
only noticed the headings with kdiff3, looked right past them before that.

yogo1212 commented:
here are the amended changes:
```
diff --git a/target/linux/ath79/dts/qca9563_compex_wpj563.dts b/target/linux/ath79/dts/qca9563_compex_wpj563.dts
index 75ab92e6a4..ae9dd5c460 100644
--- a/target/linux/ath79/dts/qca9563_compex_wpj563.dts
+++ b/target/linux/ath79/dts/qca9563_compex_wpj563.dts
@@ -10,6 +10,19 @@
 	model = "Compex WPJ563";
 	compatible = "compex,wpj563", "qca,qca9563";
 
+	aliases {
+		label-mac-device = &eth0;
+		led-boot = &led_sig4;
+		led-failsafe = &led_sig4;
+		led-running = &led_sig4;
+		led-upgrade = &led_sig4;
+	};
+
+	beeper {
+		compatible = "gpio-beeper";
+		gpios = <&gpio 19 GPIO_ACTIVE_HIGH>;
+	};
+
 	keys {
 		compatible = "gpio-keys";
 
@@ -39,7 +52,7 @@
 			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
 		};
 
-		sig4 {
+		led_sig4: sig4 {
 			label = "wpj563:green:sig4";
 			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
 		};
@@ -90,6 +103,7 @@
 	status = "okay";
 
 	phy-mask = <0>;
+
 	phy0: ethernet-phy@0 {
 		reg = <0>;
 		phy-mode = "sgmii";
@@ -128,5 +142,4 @@
 	status = "okay";
 
 	mtd-cal-data = <&art 0x1000>;
-	mtd-mac-address = <&uboot 0x18>;
 };

```

adrianschmutzler commented:
Thanks, I'm fine with the latest version and will merge it.

Apart from that, I realized for the first time that all WPJ devices just have a USB header. So, is USB used for anything _by default_? Otherwise, I'd remove the USB packages from DEVICE_PACKAGES for all three devices as they do not appear to be used in default setup ...

yogo1212 commented:
@adrianschmutzler there's no intended purpose for the USB port that i know of.
in my view, the mpci slot and the USB header are comparable in that regard.
if find adding, for example, an additional radio to any of those three boards just a bit easier [through USB](https://www.aliexpress.com/item/32996684870.html) than through the miniPCIe module.

from a code perspective, it makes to be critical of the potentially unused dependency.
from a user perspective, it might be confusing to have the miniPCIe interface ready to go but not the USB header.
on the other hand, it's easy enough to get the USB kmods through opkg.

thanks for your time and be well :-)
